### ISSUE TITLE
Prometheus Agent Support

### ISSUE THREAD

**Issue by kakkoyun** [prometheus-operator/prometheus-operator#3989]
Prometheus soon is going to have a lightweight version as an agent. https://github.com/prometheus/prometheus/issues/8305
It would be awesome to support this in the promethues-operator. 

I've opened the issue to start the discussion and know what the maintainers think?

@prometheus-operator/kube-prometheus-reviewers

P.S: I'd be happy to work on this.

**Comment by simonpasquier** [comment]
+1 for me definitely. Because the agent won't have all capabilities of the Prometheus server (e.g. no rule evaluation,no alertmanager config, no API probably) and the plan is to have it as another binary, it might make sense to have another CRD for it. Obviously we need more details on what the agent will actually look like.

**Comment by brancz** [comment]
I think it might be a little early as the details haven't been fleshed/implemented out on Prometheus side even. It doesn't hurt to think about a high-level strategy of the CRD though.

I realize that this might be overloading things a bit, but this is sort of an opportunity for a new start, where we could re-think our mistakes from the existing Prometheus CRD/API. If you have time, I would very much welcome it if we could think about the ScrapeConfig CRD as the basis of all configuration, and PodMonitors/ServiceMonitors/Probes only compile down to the ScrapeConfig CRD and all the PrometheusAgent CRD does is pick up ScrapeConfig CRDs directly. Rough thoughts here: https://github.com/prometheus-operator/prometheus-operator/issues/2787

**Comment by ArthurSens** [comment]
This issue came up during the contributor office hours, so just backporting the discussion here:

It’s already possible to run the agent once it's released with the Prometheus CR so it's not like we have a huge urgency on this, but a dedicated CR for it would definitely help with validation and a better UX.

* The idea for the new Agent CR is to re-use everything we can from the Prometheus CR. 
* Creating a new Agent CR will not prevent people from using Prometheus CR with the agent configuration, and we should not block this from happening. The main reason is because avoiding maintaining extra logic for little benefit

**Comment by simonpasquier** [comment]
Now that we have a beta release of Prometheus with the agent feature flag ([v2.32.0-beta.0](https://github.com/prometheus/prometheus/releases/tag/v2.32.0-beta.0)), I gave it a try using the existing Prometheus CRD.

TL;DR: it's not working out of the box but it shouldn't be too hard to adjust a bit the operator to "support" the agent mode with the existing Prometheus CR. 

My CR looks like this:

```
apiVersion: monitoring.coreos.com/v1
kind: Prometheus
metadata:
  labels:
    app.kubernetes.io/name: prometheus
    prometheus: agent
  name: agent
  namespace: openshift-monitoring
spec:
  image: quay.io/prometheus/prometheus:v2.32.0-beta.0
  containers:
  - name: prometheus
    args:
    - '--config.file=/etc/prometheus/config_out/prometheus.env.yaml'
    - '--storage.agent.path=/prometheus'
    - '--enable-feature=agent'
    - '--web.enable-lifecycle'
  - url: http://prometheus-receiver:9090/api/v1/write
    remoteTimeout: "5s"
  replicas: 1
```

The important bit is that the Prometheus arguments need to be overridden because the storage flags are different (TSDB path and retention in particular). I've used another Prometheux server with `--enable-feature=remote-write-receiver` as the remote write receiver.

The blocking issue right now is that the generated configuration isn't compatible with the agent mode because the operator generates the `rules` and `alerting` sections which isn't permitted for the agent. As a workaround, I've set the resource in paused mode, modified the generated secret manually to remove the 2 keys and it worked!
I think that it would be fine to tweak the generated Prometheus configuration so that `rules` and `alerting` fields aren't present when the custom resource has nil rule selectors and alerting spec.

**Comment by sherifkayad** [comment]
That will be really great to have the Agent feature. We are much looking forward for it

**Comment by simonpasquier** [comment]
@fpetkovski started https://docs.google.com/document/d/1LnnxGOaypmwgczuTp_nLyWG0EiF6ohNCsZf3XQHDvkc/edit to kick off the discussion :)

**Comment by sherifkayad** [comment]
@simonpasquier is there a rough timeline on when the Agent CRD support could be started / completed?

**Comment by simonpasquier** [comment]
@sherifkayad hard to commit on something but @fpetkovski has submitted #4512 to discuss the design and feedback is very welcome.

**Comment by tanvp112** [comment]
> I've set the resource in paused mode, modified the generated secret manually to remove the 2 keys and it worked!

Can you help to explain how to set to pause mode and remove the two fields? As far as I understood these fields are control by the operator. Thanks!

**Comment by ElkRom** [comment]
Hi,
When this issue will be fixed?

**Comment by simonpasquier** [comment]
@ElkRom as far I know, we don't have an ETA.

**Comment by avdicl** [comment]
Keeping this issue alive.

**Comment by pavan541cs** [comment]
@simonpasquier If the design is inplace and the team requires hands for development ... I can help out here if needed ...

**Comment by fpetkovski** [comment]
@pavan541cs I will not have time in the following weeks to work on this, so any help from your side would be highly appreciated. If you want to start working on this, I can help out with reviewing your code or answering any questions.

**Comment by pavan541cs** [comment]
@fpetkovski Is there any design document inplace to work on this or should I start from scratch by designing and then implementing the same ?

**Comment by fpetkovski** [comment]
Yes, we have this document that you can use as a base: https://github.com/prometheus-operator/prometheus-operator/blob/main/Documentation/designs/prometheus-agent.md

**Comment by simonpasquier** [comment]
I know that @paramite also expressed interest in working on this feature. Just make sure that both of you don't work in parallel :)

**Comment by paramite** [comment]
That's true. In case there's nobody who already started working on this, I can start with implementation.

**Comment by pavan541cs** [comment]
@paramite I have not started the implementation yet ... We want to enable agent mode for Prometheus on a very urgent basis ... Let me know when you can complete the implementation and also if you require an additional helping hand

**Comment by simonpasquier** [comment]
@pavan541cs note that if you follow https://github.com/prometheus-operator/prometheus-operator/issues/3989#issuecomment-974137486, it's possible to deploy a Prometheus object in agent mode already (the only "hack" is that you need to repeat the complete `args` list).

**Comment by ArthurSens** [comment]
Hi @simonpasquier, during today's office hours we had a brief discussion about the implementation of the PrometheusAgent CRD. The discussion was focuses on whether we would create a whole new package/controller for prometheusagent or if we would have a shared controller between Prometheus and PrometheusAgent.

By having a shared controller we would avoid a lot of code deduplication, but that would require a big refactoring to generalize for both CRDs. With a dedicated package/controller we would basically change the order here, we can go straight for the implementation and have something working fast, but a refactor would be useful to reduce a lot of code that would be duplicated.

We would like to get your opinion on this one. Do you already have an idea about how the implementation should be done?

**Comment by ArthurSens** [comment]
Either way, I think that generalizing whatever piece of code that uses `CommonPrometheusFields`, e.g. config generation, is a good step forward.

Do you think I could start working on this type of generalization already?

**Comment by simonpasquier** [comment]
hey @ArthurSens! My original idea was indeed to work on factoring out the config generation code to make it agnostic of the [Prometheus](https://github.com/prometheus-operator/prometheus-operator/blob/86336af6b3f71bdca225b463a4daa88b1985865a/pkg/apis/monitoring/v1/prometheus_types.go#L352-L362) struct as I felt that it was the easier part to get started and where most of the duplication would be.

E.g. instead of this https://github.com/prometheus-operator/prometheus-operator/blob/86336af6b3f71bdca225b463a4daa88b1985865a/pkg/prometheus/promcfg.go#L373

do this
```golang
func buildExternalLabels(p v1.CommonPrometheusFields) yaml.MapSlice { 
```

**Comment by simonpasquier** [comment]
Closed by #5310. The PrometheusAgent CRD is available starting with v0.64.0.

Awesome work by @ArthurSens :partying_face:
